### PR TITLE
Setup unit testing for Table InlineText/BodyRow Component

### PR DESCRIPTION
**Pull request by Marswang92**
# Description
- This story will be to set up unit tests for the shared Table component, following the pattern and interface for running tests from within particles.
- Only invoke `composeStyles` when necessary in `InlineText`.

# Tasks
- [x] Install dependencies for testing.
- [x] Unit Testing for `InlineText`.
- [x] Unit Testing for `BodyRow`.

~~- [ ] Unit Testing for `InlineSelect`.~~
~~- [ ] Unit Testing for `InlineDateSelect`.~~
~~- [ ] Unit Testing for `GroupRow`.~~
~~- [ ] Unit Testing for `BodyRows`.~~
~~- [ ] Unit Testing for `Caret`.~~
~~- [ ] Unit Testing for `HeadCell`.~~
~~- [ ] Unit Testing for `HeadRow`.~~
~~- [ ] Unit Testing for `SideBar`.~~
~~- [ ] Unit Testing for `Table`.~~
~~- [ ] (?) Snapshot testing.~~

### CODE REVIEW

File: table/src/InlineText.jsx [hunk 1]
```diff
@@ -43,6 +38,12 @@ export default class InlineText extends React.Component {
     }
   }
 
+  componentWillReceiveProps(nextProps) {
+    if (this.props.value !== nextProps.value) {
+      this.setState({ value: nextProps.value });
+    }
+  }
```
Comment (ben-burlingham) [inline table/src/InlineText.jsx]: Merge master should fix these

File: table/package.json [hunk 2]
```diff
@@ -6,7 +6,13 @@
   "scripts": {
     "dev": "NODE_ENV='development' webpack --progress --colors --watch",
     "prod": "NODE_ENV='production' webpack --progress --colors -p",
-    "sandbox": "NODE_ENV='development' webpack-dev-server --progress --colors --hot --inline ./sandbox/sandbox.jsx --config webpack.dev.config.js  --content-base ./../sandbox"
+    "sandbox": "NODE_ENV='development' webpack-dev-server --progress --colors --hot --inline ./sandbox/sandbox.jsx --config webpack.dev.config.js  --content-base ./../sandbox",
+    "test": "jest --runInBand --verbose"
```
Comment (mcclayton) [inline table/package.json]: Any reason we want to disable running tests in parallel via `--runInBand`?

File: table/test/InlineSelect.test.jsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  );
+};
+
+let inlineTextWrapper;
+
+beforeEach(() => {
+  inlineTextWrapper = renderInlineEdit();
+});
+
+describe('', () => {
```
Comment (mcclayton) [inline table/test/InlineSelect.test.jsx]: Typically the outermost `describe` would be the name of the component being tested. i.e. `describe('InlineSelect', ...`

File: table/test/InlineSelect.test.jsx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  describe('Constructor', () => {
+    it('initializes `value` state with `value.` prop', () => {
+      expect(inlineTextWrapper.state('value')).toEqual(inlineTextProps.value);
+    });
+  });
+  describe('Input', () => {
+    it('should show after clicking on the container.', () => {
```
Comment (mcclayton) [inline table/test/InlineSelect.test.jsx]: The `it` statement here doesn't really describe well what the test is testing.
Perhaps this should be 
```
describe('Input Container', () => {
    describe('onClick', () => {
        it('should set `editing` state value to `true`.', () => {
```
Perhaps there is a better way to word the describes - but hopefully you get the idea.

File: table/test/InlineSelect.test.jsx [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      inlineTextWrapper.find('div').at(0).simulate('click');
```
Comment (mcclayton) [inline table/test/InlineSelect.test.jsx]: This is a pretty brittle selector. Can you use http://airbnb.io/enzyme/docs/api/ReactWrapper/ref.html to select this div via it's `ref` ?

File: table/test/InlineSelect.test.jsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      expect(inlineTextWrapper.state('editing')).toEqual(true);
+    });
+    it('should show `textEmpty` props if no `value` provided.', () => {
+      const wrapperWithoutValue = renderInlineEdit({ value: '' });
+      expect(wrapperWithoutValue.find('div').at(2).text()).toEqual(inlineTextProps.textEmpty);
```
Comment (mcclayton) [inline table/test/InlineSelect.test.jsx]: Again, is there a better selector you can use here? Perhaps by a className, ID, ref?
I try to avoid selectors that rely on ordering at all costs.

File: table/test/InlineSelect.test.jsx [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    });
+    it('should update `value` state correctly on change.', () => {
+      inlineTextWrapper.find('div').at(0).simulate('click');
```
Comment (mcclayton) [inline table/test/InlineSelect.test.jsx]: Better selector?

Comment (mcclayton) [inline table/test/InlineSelect.test.jsx]: What does the click do? Is this to set some sort of state variable before testing the changeEvent?
It's best not to have other functionality in a unit test that is unrelated to what it is you are purely trying to unit test. If this is just to get the input to show up - it'd be better to manually set the state of the component before the test rather than triggering a click event to do this.

File: table/test/InlineSelect.test.jsx [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      const inlineTextInput = inlineTextWrapper.find('input');
+      const expectedChangeValue = 'This is a changed search string';
+      const mockChangeEvent = {
+        target: { value: expectedChangeValue }
+      };
+      inlineTextInput.simulate('change', mockChangeEvent);
+      expect(inlineTextWrapper.state('value')).toEqual(expectedChangeValue);
+    });
+    it('should clear `value` state on click clear button.', () => {
+      inlineTextWrapper.find('button').simulate('click');
```
Comment (mcclayton) [inline table/test/InlineSelect.test.jsx]: Button is "ok" here, but this could still be better. The moment another button is added to this component, this selector could potentially fail the test when it shouldn't.

File: table/test/InlineSelect.test.jsx [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      expect(inlineTextWrapper.state('value')).toEqual('');
+    });
+    it('should call appropriate onSubmit handler on Enter key up with `value` state if `value` has been updated.', () => {
```
Comment (mcclayton) [inline table/test/InlineSelect.test.jsx]: This test is doing a lot here - It would be better if you were able to manually set the `value` in the state via: http://airbnb.io/enzyme/docs/api/ReactWrapper/setState.html instead of relying on the change event to do this. This keeps your test purely a unit test by only testing a single unit of code per test.

Also, I think the `describes` and `it` statement can then be cleaned up to something like:
```
describe('InlineSelect', () => {
    describe('Input', () => {
        describe('On Enter Key Up', () => {
             it('should call appropriate onSubmit handler with appropriate `value`.', () => {
```

**Comment by mcclayton** [comment]
**Thank you so much** for taking the time to write unit tests. This is seriously incredibly needed!
You rock! 🚀  💯

File: table/package.json (hunk 2 above)
Comment (Marswang92) [reply table/package.json]: Oh previously I was trying to solve the issue of console log not being emitted in the terminal. Will remove it.

File: table/test/InlineSelect.test.jsx (hunk 7 above)
Comment (Marswang92) [reply table/test/InlineSelect.test.jsx]: OK gotcha, will do a `setState` instead of trying to mock a click!

File: table/test/InlineSelect.test.jsx (hunk 8 above)
Comment (Marswang92) [reply table/test/InlineSelect.test.jsx]: Yeah actully wanna discuss with you about some issue of using class selector with css-module

File: table/test/data/groupedData.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
+module.exports = [
```
Comment (mcclayton) [inline table/test/data/groupedData.js]: Typically a test directory like this is called `/fixtures`.

File: table/test/BodyRow.test.jsx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    });
+    it('should be checked if prop `isChecked` is set to true.', () => {
+      expect(bodyRowWrapper.find('input').prop('checked'))
+        .toEqual(props.isChecked);
+    });
+    describe('On Change', () => {
+      it('should call appropriate onCheck handler with row key.', () => {
+        const onCheckSpy = sinon.spy();
+        const wrapperWithSpy = renderBodyRow({ onCheck: onCheckSpy });
+        const mockChangeEvent = {
```
Comment (mcclayton) [inline table/test/BodyRow.test.jsx]: You don't need the `mockChangeEvent` here do you?
`onChange` does not doing anything with the actual event so you can simply call `wrapperWithSpy.find('input').simulate('change')`.

File: table/test/BodyRow.test.jsx [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  describe('CheckBox', () => {
+    it('should be rendered if prop `showCheckboxes` is set to true.', () => {
+      expect(bodyRowWrapper.find(`.${styles.controlCheckbox}`).exists()).toEqual(true);
+    });
+    it('should be checked if prop `isChecked` is set to true.', () => {
+      expect(bodyRowWrapper.find('input').prop('checked'))
+        .toEqual(props.isChecked);
+    });
+    describe('On Change', () => {
+      it('should call appropriate onCheck handler with row key.', () => {
```
Comment (mcclayton) [inline table/test/BodyRow.test.jsx]: Did you mean appropriate `row key value` maybe?

**Comment by mcclayton** [comment]
To keep things smaller, I'd perhaps isolate this PR to just the `InlineText` and `BodyRow` components. You can add others in individual PRs :)

File: table/src/InlineText.jsx [hunk 13]
```diff
@@ -10,36 +10,41 @@ export default class InlineText extends React.Component {
...
+      value: props.value
+    };
   }
 
-  componentWillReceiveProps(nextProps) {
-    if(this.props.value !== nextProps.value) {
-      this.setState({ value: nextProps.value })
+  componentDidMount() {
+    // Only do this in browser.
+    if (this.container.parentNode.offsetWidth < 75 && !jest) {
```
[the remark this answers is not in the document]
Comment (Marswang92) [reply table/src/InlineText.jsx]: Doing this cuz the warning is bit annoying in testing, thought it should only be triggered while it's being rendered in browser. (btw not a big fan of the parent Node Idea, cuz in different use case the `InlineText`'s parentNode might vary).

[the remark this answers is not in the document]
Comment (Marswang92) [reply table/src/InlineText.jsx]: Test env is a better idea or I can just mock the container, will think about it!

File: table/test/data/groupedData.js (hunk 10 above)
Comment (Marswang92) [reply table/test/data/groupedData.js]: gotcha thanks!

File: table/test/BodyRow.test.jsx (hunk 11 above)
Comment (Marswang92) [reply table/test/BodyRow.test.jsx]: Yes haha removing it!

File: table/test/BodyRow.test.jsx [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+      expect(bodyRowWrapper.find(`.${styles.tableCell}`).length).toEqual(2);
+    });
+  });
+  describe('CheckBox', () => {
+    it('should be rendered if prop `showCheckboxes` is set to true.', () => {
+      expect(bodyRowWrapper.find(`.${styles.controlCheckbox}`).exists()).toEqual(true);
+    });
+    it('should be checked if prop `isChecked` is set to true.', () => {
+      expect(bodyRowWrapper.find('input').prop('checked'))
+        .toEqual(props.isChecked);
```
[the remark this answers is not in the document]
Comment (Marswang92) [reply table/test/BodyRow.test.jsx]: Yeah i think in this case the better way to do is instead using `bodyRowWrapper`, just make a new wrapper explicitely set the prop like `const checkedWrapper = renderBodyRowWrapper({ isChecked: true })`?

File: table/test/BodyRow.test.jsx [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+let bodyRowWrapper;
+
+beforeEach(() => {
+  bodyRowWrapper = renderBodyRow();
+});
+
+describe('BodyRow', () => {
+  describe('Cells', () => {
+    it('should hide if hidden in definition is set to true.', () => {
+      expect(bodyRowWrapper.find(`.${styles.tableCell}`).length).toEqual(2);
```
[the remark this answers is not in the document]
Comment (Marswang92) [reply table/test/BodyRow.test.jsx]: So if the `def` set to be hidden the cell will be null. So here is an `existence` check, though it should be more specific e.g. instead of checking the length of nodes, check the existence of a specific node with def hidden.

File: table/src/InlineText.jsx [hunk 16]
```diff
@@ -10,36 +10,41 @@ export default class InlineText extends React.Component {
...
+    textEmpty: PropTypes.string,
     value: PropTypes.string
   };
 
   static defaultProps = {
     canClear: true,
     onClear: () => {},
-    stylesheets: [],
-    value: 'Input...'
+    stylesheets: undefined,
```
[the remark this answers is not in the document]
Comment (Marswang92) [reply table/src/InlineText.jsx]: Doing this here is just to shut eslint off and to avoid calling `composeStyles` when `stylesheets` is not provided. And yes actually set it default to be an array and check isEmpty in the constructor is a better idea. Thanks!

**Comment by Marswang92** [comment]
@mcclayton For sure, will update this pr accordingly! Thanks a lot for all the super helpful advice!

**Comment by ben-burlingham** [comment]
Is there a reason the fixtures reproduce the test data? Can they just pull from the sandbox test data?

**Comment by Marswang92** [comment]
@ben-burlingham for sure we can. Doing this cuz lot of times we change stuff in the sandbox data for developing new features. The fixture folder instead should keep the same?

**Comment by ben-burlingham** [comment]
I'd say exactly the opposite - the data used in the sandbox _is_ test data, and is there intentionally to test empty cases, overflow cases, etc.

Also, we don't want to maintain two stubs ;)

File: table/test/BodyRow.test.jsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (mcclayton) [reply table/test/BodyRow.test.jsx]: @Marswang92 Yes I like that option much better 👍

**Comment by mcclayton** [comment]
Hmmm, I think @Marswang92 has a point that I agree with:
My only hesitation to only trying to use a single fixture for both is that while fixtures used for the sandbox could make good tests in the unit tests' suite, the reverse is not necessarily true. I'd hate to preclude ourselves from valuable test case fixtures simply because they don't make sense in the context of the sandbox. Anywho, just my two cents (I'm fine with whatever)
¯\_(ツ)_/¯

File: table/test/BodyRow.test.jsx [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+beforeEach(() => {
+  bodyRowWrapper = renderBodyRow();
+});
+
+describe('BodyRow', () => {
+  describe('Cells', () => {
+    it('should not be rendered if hidden in definition is set to true.', () => {
+      expect(bodyRowWrapper.find(`.${styles.tableCell}`).length)
+        .toEqual(columnDefinitions.filter(o => o.hidden !== true).length);
```
Comment (mcclayton) [inline table/test/BodyRow.test.jsx]: Nice, this is much better 👍

Comment (mcclayton) [inline table/test/BodyRow.test.jsx]: Also, back to the fixtures, sometimes it's better to inline them, or split specific fixtures out into their respective cases such as:
```
/fixtures/hiddenColumns.json
```
That way, you don't have one large shared fixture that is prone to breaking multiple tests when you modify it. Better to tailor your data to individual tests when possible vs. attempting to tailor one large fixture to the entire test suite.
I think this is fine for now, but just for future reference.

**Comment by ben-burlingham** [comment]
@mcclayton you're the lead on the testing initiative, let's do what you think is best :)

**Comment by mcclayton** [comment]
We're all "leads" ❤️

File: table/test/BodyRow.test.jsx (hunk 17 above)
Comment (Marswang92) [reply table/test/BodyRow.test.jsx]: Got it this is very helpful. I'm on the way!
